### PR TITLE
fix: Cleanup tags when component is removed

### PR DESCRIPTION
**Pull request by jithumkunjuraman**
To be frugal with the tag quotas imposed by GCP we should clean up any tag values and keys that are unused, e.g. because all related resources were deleted

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/onsi/ginkgo/v2 v2.17.1
 	github.com/onsi/gomega v1.32.0
 	github.com/patrickmn/go-cache v2.1.0+incompatible
+	github.com/rs/zerolog v1.33.0
```
Comment (mKeRix) [inline go.mod:13]: This dependency doesn't seem like it is used in the current branch code - maybe run `go mod tidy` again? At least on my machine that dropped it from the dependencies.

File: internal/gcp/tags.go [hunk 2]
```diff
@@ -178,3 +180,45 @@ func (m *tagsManager) GetProjectInfo(ctx context.Context, projectID string) (*re
...
+
+func (m *tagsManager) DeleteValue(ctx context.Context, projectID string, key string, value string) error {
+
+	req := &resourcemanagerpb.DeleteTagValueRequest{
+		Name: value,
+	}
+
+	op, err := m.valuesClient.DeleteTagValue(ctx, req)
+	if err != nil {
+		return nil // Tag value already in use or deleted, consider this a success
```
Comment (mKeRix) [inline internal/gcp/tags.go:192]: Is there any way to differentiate in use/deleted errors from e.g. permission errors? The [API reference](https://cloud.google.com/resource-manager/docs/core_errors#Global_Errors) suggests that the API utilizes multiple error types that might help differentiate, but I'm not sure what your tests against the real API showed.

File: internal/gcp/tags.go [hunk 3]
```diff
@@ -178,3 +180,45 @@ func (m *tagsManager) GetProjectInfo(ctx context.Context, projectID string) (*re
...
+	}
+	_, err = op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to delete the tagValue %w", err)
+	}
+
+	m.cache.Delete(cacheKeyTagValue(key, value))
```
Comment (mKeRix) [inline internal/gcp/tags.go:199]: I wonder what happens if the tag value is in the delete operation and then a new resources comes in that attaches that exact tag value. That might be quite difficult to test though, so I'm happy to just accept this as an edge case that we don't plan for (yet).

File: internal/gcp/tags.go [hunk 4]
```diff
@@ -178,3 +180,45 @@ func (m *tagsManager) GetProjectInfo(ctx context.Context, projectID string) (*re
 	m.cache.Set(cacheKey, project, tagCacheDuration)
 	return project, nil
 }
+
+func (m *tagsManager) DeleteValue(ctx context.Context, projectID string, key string, value string) error {
```
Comment (mKeRix) [inline internal/gcp/tags.go:184]: Should this maybe also be named `DeleteValueIfUnused` like the other function for consistency?

**Review by mKeRix — COMMENTED** [review]
Left some comments, but the overall logic looks good to me!

File: go.mod (hunk 1 above)
Comment (jithumkunjuraman) [reply go.mod:13]: Will remove. Added for debugging

File: internal/gcp/tags.go (hunk 3 above)
Comment (jithumkunjuraman) [reply internal/gcp/tags.go:199]: We have tested this. The new tag key/value gets added but with a minimal delay.

File: internal/gcp/tags.go (hunk 4 above)
Comment (jithumkunjuraman) [reply internal/gcp/tags.go:184]: Sure makes sense, I will change the name

File: internal/gcp/tags.go (hunk 2 above)
Comment (jithumkunjuraman) [reply internal/gcp/tags.go]: Added the error check

**Review by mKeRix — APPROVED** [review]
LGTM, nice work! 🚀
